### PR TITLE
build and test Bitcoin Core with the nix-shell

### LINKED ISSUE #2: have a CI job that tests the shell/environment by building Bitcoin Core
possibly a testing matrix with different nixos versions, compilers (see #1), ...

### PR DESCRIPTION
**Pull request by 0xB10C**
A basic CI job that clones different Bitcoin Core versions and builds them and runs the tests using the nix-shell.

Closes #2

### CODE REVIEW

File: .github/workflows/nix-shell-build.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+      with:
+        nix_path: nixpkgs=channel:${{ matrix.nixos-version }}
+    - name: Checkout Bitcoin Core
+      uses: actions/checkout@v4
+      with:
+        repository: bitcoin/bitcoin
+        ref: ${{ matrix.bitcoin-ref }}
+        path: bitcoin
+    - name: Build with nix-shell (and aliases)
+      run: nix-shell --command "cd bitcoin && bash -c 'a && c && m && t'"
```
Comment (0xB10C) [inline .github/workflows/nix-shell-build.yml]: hm yeah that won't work: `bash: line 1: a: command not found`

I've also tried without invoking bash - same problem.

see also: https://github.com/NixOS/nix/issues/1598

Comment (0xB10C) [reply .github/workflows/nix-shell-build.yml]: Bash aliases in non-interactive shells seem generally to be a problem, even when setting `shopt -s expand_aliases`. Using bash functions should work both in interactive and non-interactive shells.

**Comment by 0xB10C** [comment]
Looking at https://github.com/cbergqvist/bitcoin-core-overlay/blob/7731950823f7b9594b0e4bf200270696eb0d8d73/shell.nix#L178C7-L178C28

there might be cases where using an alias is preferable to a function (but I think both ways work). For e.g. `bitcoin-cli`, non-interactive use is unlikely so an alias should be fine too.

File: .github/workflows/nix-shell-build.yml (hunk 1 above)
Comment (cbergqvist) [reply .github/workflows/nix-shell-build.yml]: Irritating that it seems to be working for some but not for others in the referenced nix issue. Feels a bit cleaner with aliases but I can live with the current compromise here for now. Thanks for trying to make it work.

File: shell.nix [hunk 2]
```diff
@@ -83,38 +83,32 @@ in pkgs.mkShell {
...
-      alias ft="python3 test/functional/test_runner.py"
-      # all tests
-      alias t="ut && ft"
-
+      # Creates a new shorthand with the name $1 aliasing to the command $2
+      # We don't use aliases as they might not work in non-interactive shell sessions
+      # such as a CI. 
+      shorthand() {
+        eval "$1() { bash -c '$2'; }"
+        echo "alias $1 = $2"
```
Comment (cbergqvist) [inline shell.nix]: Ah, was surprised that it was still outputting "alias $1 = $2". Feels a bit fake, one can no longer do:
```console
$ alias t
bash: alias: t: not found
```

Feels more honest to strip away the alias and output a heading before:
```
...
adding bcc egg to PYTHONPATH: /nix/store/b41np9nwysnh2s4jdsxd4jywhjm4myz4-bcc-0.30.0/lib/python3.11/site-packages/bcc-0.30.0-py3.11.egg

Shorthands:
a = sh autogen.sh
c = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR
c_no-wallet = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR --disable-wallet
...
```

Comment (0xB10C) [reply shell.nix]: Good point. Changed it to:

```
...
Shorthands:
  a = sh autogen.sh
  c = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR
  c_no-wallet = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR --disable-wallet
  c_fast = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR --disable-wallet --disable-tests --disable-fuzz --disable-bench -disable-fuzz-binary
  c_fast_wallet = ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR --disable-tests --disable-bench
  m = make -j8
  cm = c & m
  cm_fast = c_fast && m
  acm = a && c && m
  acm_nw = a && c_no-wallet && m
  acm_fast = a && c_fast && m
  acm_fast_wallet = a && c_fast_wallet && m
  ut = make check
  ft = python3 test/functional/test_runner.py
  t = ut && ft
...
```

**Review by cbergqvist — APPROVED** [review]
Nice!

**Comment by cbergqvist** [comment]
Only actually tested the shorthand changes now and it appears the indirect ones aren't working:
```console
$ acm
bash: line 1: a: command not found
```

**Comment by 0xB10C** [comment]
> Only actually tested the shorthand changes now and it appears the indirect ones aren't working:
> 
> ```
> $ acm
> bash: line 1: a: command not found
> ```

Hmm, yes. I can reproduce it..

One possibility would be to explicitly define e.g. `acm` to be `sh autogen.sh && ./configure --with-boost-libdir=$NIX_BOOST_LIB_DIR && make -jX`...

**Comment by cbergqvist** [comment]
This works on the user CLI, not sure how it fares on CI:
```bash
      shorthand() {
        eval "$1() { bash -c '$2'; }"
        echo "  $1 = $2"
        export SH_$1="$2"
      };

      echo "Shorthands:"
      shorthand "a"               "sh autogen.sh"
      shorthand "c"               "./configure --with-boost-libdir=\$NIX_BOOST_LIB_DIR"
      shorthand "c_no_wallet"     "./configure --with-boost-libdir=\$NIX_BOOST_LIB_DIR --disable-wallet"
      shorthand "c_fast"          "./configure --with-boost-libdir=\$NIX_BOOST_LIB_DIR --disable-wallet --disable-tests --disable-fuzz --disable-bench -disable-fuzz-binary"
      shorthand "c_fast_wallet"   "./configure --with-boost-libdir=\$NIX_BOOST_LIB_DIR --disable-tests --disable-bench"
      shorthand "m"               "make -j${jobs}"
      shorthand "cm"              "$SH_c && $SH_m"
      shorthand "cm_fast"         "$SH_c_fast && $SH_m"
      shorthand "acm"             "$SH_a && $SH_c && $SH_m"
      shorthand "acm_nw"          "$SH_a && $SH_c_no-wallet && $SH_m"
      shorthand "acm_fast"        "$SH_a && $SH_c_fast && $SH_m"
      shorthand "acm_fast_wallet" "$SH_a && $SH_c_fast_wallet && $SH_m"
      shorthand "ut"              "make check" 
      shorthand "ft"              "python3 test/functional/test_runner.py"
      shorthand "t"               "$SH_ut && $SH_ft"
```

**Comment by cbergqvist** [comment]
One further issue with the current `shorthand()` function declaration implementation is that extra parameters get ignored.

Previously using `alias` one could type:
```$ c --help```
...and get the expected result. Now `c` will be executed by itself and ` --help` will be ignored. Maybe `shorthand()` can be fixed to declare functions that take optional parameters.

**Comment by cbergqvist** [comment]
At this point it might be better to revert back to using `alias` as a base + do something like `shorthand()` only for those ones called by CI. What do you think?

**Comment by 0xB10C** [comment]
I agree that the current situation isn't optimal and we can improve on it. The idea behind having the CI use the the alias/shorthand was to have "test coverage" for them. Ideally, we find a way of using the aliases in CI. I won't be able to work on it myself for the next couple of days, but I can have a look at a PR if you happen to have time to experiment.

**Pull request merged**